rotkehlchen/chain/ethereum/modules/oneinch/v4/decoder.py

```diff
@@ -0,0 +1,111 @@
+from typing import TYPE_CHECKING, Callable
+
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.chain.evm.decoding.oneinch.decoder import OneinchCommonDecoder
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.accounting.structures.evm_event import EvmEvent
+
+from rotkehlchen.chain.ethereum.modules.balancer.v2.decoder import (
+    V2_SWAP as BALANCER_V2_SWAP_SIGNATURE,
+)
+from rotkehlchen.chain.ethereum.modules.uniswap.v2.decoder import (
+    SWAP_SIGNATURE as UNISWAP_V2_SWAP_SIGNATURE,
+)
+from rotkehlchen.chain.ethereum.modules.uniswap.v3.decoder import (
+    SWAP_SIGNATURE as UNISWAP_V3_SWAP_SIGNATURE,
+)
+
+from ..constants import CPT_ONEINCH_V4
+from .constants import ONEINCH_V4_MAINNET_ROUTER
+
+
+class Oneinchv4Decoder(OneinchCommonDecoder):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+            router_address=ONEINCH_V4_MAINNET_ROUTER,
+            swapped_signatures=[
+                UNISWAP_V3_SWAP_SIGNATURE,
+                UNISWAP_V2_SWAP_SIGNATURE,  # uniswap v2 is also used by sushiswap
+                BALANCER_V2_SWAP_SIGNATURE,
+            ],
+            counterparty=CPT_ONEINCH_V4,
+        )
+
+    def _handle_post_decoding(
+            self,
+            transaction: EvmTransaction,
+            decoded_events: list['EvmEvent'],
+            all_logs: list[EvmTxReceiptLog],
+    ) -> list['EvmEvent']:
+        """
+        It handles the case where a swap is done via 1inch v4 on any of the known pool contracts.
+        In that case there are 3 simple decoded events already created: Fee, spend and receive.
+        If Spend and receive events are not decoded by the swap decoder (for example from uniswap
+        v3 decoder), they should be decoded by this aggregator decoder.
+        The decoding is done with this post decoding rule and not from
+        addresses_to_decoders mapping because the 1inch v4 router is not in the addresses of the
+        logs of the transaction, and consequently it can't be extracted and mapped.
+        """
+        for tx_log in all_logs:
+            if tx_log.topics[0] in self.swapped_signatures:
```

Comment:
Hmm then this is not needed. Why are you runnign this rule? There is no point. Just to change counterparty and notes from uniswa/sushiswap to 1inch?

This makes no sense. Those transactions can be left decoded as they were. What you need to do is decode transactions that are not decoded properly at all and are using the aggregator part of 1inch.